From 9f06f9d19cd55f6fe6116b859b74840611d43ab6 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 13 Mar 2025 10:26:34 +0800
Subject: [PATCH 154/254] mtk: hostapd: refactor MTK configuration options in
 hostapd/wpa_s for prplMesh

There are hostapd/wpa_s options for MLD operation in MTK internal patch,
but prplMesh will not use these options.
Therefore this commit refactor the usage of these option by following

- mld_primary: it is only used by ucode, so remove it from hostapd
- mld_allowed_links: it is needed in AP MLD mode for starting beaconing
  after AP MLD is ready, and in Extender mode for STA interface checking
  whether AP MLD finishes the setup.
  For prplMesh use, set the default value to 255 and ignore it if the
  value is 255.
- mld_radio_msk: it is used by ucode to create wdev with correct radio
  mask. This commit makes hostapd read it from HW feature, so it can
  also be removed from hostapd configuration
- mld_allowed_phy_bitmap: set the default value to 7, which means by
  default STA MLD will use each phy radio
- mld_assoc_phy: this option remain the same. It is because the option
  is converted to "freq_list" and "mld_connect_band_pref", which are
  original wpa_s configuration and used by prplMesh

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>

Fix incorrect use on "mld_allowed_links".
1. if mld_allowed_links is 255, immediately set the beacon again
   to start beacon TX.
2. if mld_allowed_links is 0, which means the variable is not used and
   immediately set the beacon to start beacon TX.
3. Otherwise, only after the mld_allowed_links equals to valid_links,
   the beacon starts to TX.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>

Remove "mld_allowed_phy" and DEFAULT_MLD_ALLOWED_PHY.
The option 'mld_allowed_phy' is unnecessary because now UCI can set
multiple devices for a STA MLD.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ap_config.c   | 3 +++
 src/ap/hostapd.c     | 6 +++++-
 src/ap/hostapd.h     | 1 +
 src/ap/hw_features.c | 2 ++
 4 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index 68885ef9d..6e4bdb104 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -185,6 +185,9 @@ void hostapd_config_defaults_bss(struct hostapd_bss_config *bss)
 #endif
 
 	bss->bss_termination_tsf = 0;
+#ifdef CONFIG_IEEE80211BE
+	bss->mld_allowed_links = 255;
+#endif /* CONFIG_IEEE80211BE */
 }
 
 
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index bef57d4c2..b0b0caa6f 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1369,7 +1369,9 @@ static int hostapd_start_beacon(struct hostapd_data *hapd,
 	if (!conf->start_disabled && ieee802_11_set_beacon(hapd) < 0)
 		return -1;
 
-	if (hapd->conf->mld_ap && !hapd->mld->started) {
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap && (!hapd->mld->started ||
+				   hapd->conf->mld_allowed_links == 255)) {
 		struct hostapd_data *p_hapd;
 		u16 valid_links = 0;
 
@@ -1377,6 +1379,7 @@ static int hostapd_start_beacon(struct hostapd_data *hapd,
 			valid_links |= BIT(p_hapd->mld_link_id);
 
 		if (valid_links == hapd->conf->mld_allowed_links ||
+		    hapd->conf->mld_allowed_links == 255 ||
 		    !hapd->conf->mld_allowed_links) {
 			hapd->mld->started = 1;
 			ieee802_11_set_beacon(hapd);
@@ -1397,6 +1400,7 @@ static int hostapd_start_beacon(struct hostapd_data *hapd,
 		hostapd_drv_sta_deauth(hapd, addr,
 				       WLAN_REASON_PREV_AUTH_NOT_VALID);
 	}
+#endif /* CONFIG_IEEE80211BE */
 
 	if (hapd->driver && hapd->driver->set_operstate)
 		hapd->driver->set_operstate(hapd->drv_priv, 1);
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index a9dbdb4da..ff1334403 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -813,6 +813,7 @@ struct hostapd_iface {
 	struct hostapd_multi_hw_info *multi_hw_info;
 	unsigned int num_multi_hws;
 	struct hostapd_multi_hw_info *current_hw_info;
+	u8 radio_mask;
 
 #ifdef CONFIG_AFC
 	struct {
diff --git a/src/ap/hw_features.c b/src/ap/hw_features.c
index 40b0bd679..cd7c22614 100644
--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -180,12 +180,14 @@ int hostapd_get_hw_features(struct hostapd_iface *iface)
 	hostapd_free_multi_hw_info(iface->multi_hw_info);
 	iface->multi_hw_info = multi_hw_info;
 	iface->num_multi_hws = num_multi_hws;
+	iface->radio_mask = 0;
 
 	wpa_printf(MSG_DEBUG, "Multiple underlying hardwares info:");
 
 	for (k = 0; k < num_multi_hws; k++) {
 		struct hostapd_multi_hw_info *hw_info = &multi_hw_info[k];
 
+		iface->radio_mask |= BIT(hw_info->hw_idx);
 		wpa_printf(MSG_DEBUG,
 			   "  %d. hw_idx=%u, frequency range: %d-%d MHz",
 			   k + 1, hw_info->hw_idx, hw_info->start_freq,
-- 
2.45.2

